packages/structures/src/users/User.ts

```diff
@@ -0,0 +1,193 @@
+import { DiscordSnowflake } from '@sapphire/snowflake';
+import type { APIUser } from 'discord-api-types/v10';
+import { Structure } from '../Structure.js';
+import { data as kData } from '../utils/symbols.js';
+
+let UserDataTemplate: Partial<APIUser> = {};
+
+/**
+ * Sets the template used for removing data from the raw data stored for each User
+ *
+ * @param template - the template
+ */
+export function setUserDataTemplate(template: Partial<APIUser>) {
+	UserDataTemplate = template;
+}
+
+/**
+ * Gets the template used for removing data from the raw data stored for each User
+ *
+ * @returns the template
+ */
+export function getUserDataTemplate(): Readonly<Partial<APIUser>> {
+	return UserDataTemplate;
+}
+
+/**
+ * Represents any user on Discord.
+ */
+export class User<Omitted extends keyof APIUser | '' = ''> extends Structure<APIUser> {
+	public constructor(
+		/**
+		 * The raw data received from the API for the user
+		 */
+		data: Omit<APIUser, Omitted>,
+	) {
+		// Cast here so the getters can access the properties, and provide typesafety by explicitly assigning return values
+		super(data as APIUser, { template: UserDataTemplate });
+	}
+
+	public override patch(data: Partial<APIUser>) {
+		return super.patch(data, { template: UserDataTemplate });
+	}
+
+	/**
+	 * The user's id
+	 */
+	public get id() {
+		return this[kData].id as 'id' extends Omitted ? never : APIUser['id'];
+	}
+
+	/**
+	 * The username of the user
+	 */
+	public get username() {
+		return this[kData].username as 'username' extends Omitted ? never : APIUser['username'];
+	}
+
+	/**
+	 * The user's 4 digit tag, if a bot or not migrated to unique usernames
+	 */
+	public get discriminator() {
+		return this[kData].discriminator as 'discriminator' extends Omitted ? never : APIUser['discriminator'];
+	}
+
+	/**
+	 * The user's display name, the application name for bots
+	 */
+	public get globalName() {
+		return this[kData].global_name as 'global_name' extends Omitted ? never : APIUser['global_name'];
+	}
+
+	/**
+	 * The user avatar's hash
+	 */
+	public get avatar() {
+		return this[kData].avatar as 'avatar' extends Omitted ? never : APIUser['avatar'];
+	}
+
+	/**
+	 * Whether the user is a bot
+	 */
+	public get bot() {
+		return (this[kData].bot ?? false) as 'bot' extends Omitted ? never : APIUser['bot'];
```

Comment:
If we have defaults, we'll never have this return `never`, even if omitted by users

Reply:
hmm, technically we should return undefined in those cases, but there's no way for us to know that. ngl, I kind of like lying about the type since it isn't guaranteed to be accurate